File: ws-server.js
(function(){
  var Player, players, randomId, sys, ws;
  //!/usr/local/bin/coffee
  sys = require("sys");
  ws = require("./ws");
  randomId = function randomId(length) {
    var chars, id;
    chars = 'qwertyuiopasdfghjklzxcvbnm1234567890';
    id = '';
    while (id.length < length) {
      id += chars.charAt(Math.round(Math.random() * (chars.length - 1)));
    }
    return id;
  };
  Player = function Player(socket) {
    this.socket = socket;
    this.x = Math.round(Math.random() * 100);
    this.y = Math.round(Math.random() * 100);
    this.id = randomId(20);
    this.alive = true;
    return this;
  };
  Player.prototype.updatePosition = function updatePosition(x, y) {
    this.x = x;
    this.y = y;
    return this.y;
  };
  Player.prototype.kill = function kill() {
    this.alive = false;
    return this.alive;
  };
  Player.prototype.toData = function toData() {
    return "{id: '" + (this.id) + "', x: " + (this.x) + ", y: " + (this.y) + "}";
  };
  players = [];
  ws.createServer(function(websocket) {
    var player;
    player = new Player(websocket);
    websocket.addListener("connect", function(resource) {
      sys.debug(("Connected player " + (player.id)));
      websocket.write(("{yourid: '" + (player.id) + "'}"));
      players.forEach(function(p) {
        p.socket.write(player.toData());
        return websocket.write(p.toData());
      });
      websocket.write(player.toData());
      return players.push(player);
    });
    websocket.addListener("data", function(data) {
      var pos;
      pos = data.split(',');
      player.x = pos[0];
      player.y = pos[1];
      return players.forEach(function(p) {
        if (p.id !== player.id) {
          return p.socket.write(player.toData());
        }
      });
    });
    return websocket.addListener("close", function() {
      var players_buffer;
      player.kill();
      players_buffer = [];
      players.forEach(function(p) {
        if (p.alive) {
          players_buffer.push(p);
          return p.socket.write(("{removePlayer: '" + (player.id) + "'}"));
        }
      });
      players = players_buffer;
      delete player;
      return sys.debug(("Killed player " + (player.id)));
    });
  }).listen(8080);
})();

File: ws-server.rb
#!/usr/bin/ruby

require 'rubygems'
require 'em-websocket'

def random_id(length)
  chars = 'qwertyuiopasdfghjklzxcvbnm1234567890'
  id = ''

  while id.length < length
    id << chars[rand(chars.length),1]
  end
  
  id
end

class Player
  attr_reader :socket, :id, :alive
  attr_accessor :x, :y
  
  def initialize(socket)
    @socket = socket
    @x = rand(100)
    @y = rand(100)
    @id = random_id(20)
    @alive = true
  end
  
  def update_position(x, y)
    @x = x
    @y = y
  end
  
  def kill
    @alive = false
  end
  
  def to_json
    "{id: '#{@id}', x: #{@x}, y: #{@y}}"
  end
end

players = []

EventMachine::WebSocket.start(:host => "0.0.0.0", :port => 8080) do |ws|
  player = Player.new(ws)
  
  ws.onopen do
    puts "Connected player #{player.id}"
    ws.send("{yourid: '#{player.id}'}")
    
    players.each do |p|
      p.socket.send player.to_json
      ws.send p.to_json
    end
    
    ws.send player.to_json
    players << player
  end
  
  ws.onmessage do |data|
    x, y = data.split(',')
    player.update_position(x, y)
    
    players.each do |p|
      p.socket.send player.to_json unless p.id == player.id
    end
  end
  
  ws.onclose do
    player.kill
    
    players.reject! { |p| !p.alive }
    players.each do |p|
      p.socket.send "{removePlayer: '#{player.id}'}"
    end
    
    puts "Removed player #{player.id}"
    player = nil
  end
end